mdp/nodes/sfa_nodes.py

```diff
@@ -181,39 +258,108 @@ def _execute(self, x, n=None):
+        X2, _, _ = np.linalg.svd(np.dot(X1.T, np.dot(self.dcov_mtx, X1)))
+        E = np.dot(X1, X2)
+        e = np.dot(E.T, np.dot(self.dcov_mtx, E)).diagonal()
 
-        The eta value is a more intuitive measure of temporal variation,
-        defined as
-        eta(x) = t/(2*pi) * sqrt(delta(x))
-        If x is a signal of length 't' which consists of a sine function
-        that accomplishes exactly N oscillations, then eta(x)=N.
+        e = e[::-1]      # SVD delivers the eigenvalues sorted in reverse (compared to symeig). Thus
+        E = E.T[::-1].T  # we manually reverse the array/matrix storing the eigenvalues/vectors.
 
-        :Parameters:
-           t
-             Sampling frequency in Hz.
+        if rng is None:
+            self.d, self.sf = e, E
+        else:
+            self.d, self.sf = e[rng[0] - 1:rng[1]], E[:, rng[0] - 1:rng[1]]
 
-             The original definition in (Wiskott and Sejnowski, 2002)
-             is obtained for t = number of training data points, while
-             for t=1 (default), this corresponds to the beta-value defined in
-             (Berkes and Wiskott, 2005).
+    #-------------------------------------------------------------[ debugging ]
+
+    def _data_fail_msg(self):
         """
-        if self.is_training():
-            self.stop_training()
-        return self._refcast(t / (2 * numx.pi) * numx.sqrt(self.d))
+        This methods provides additional information in case _stop_training()
+        runs into a singular matrix and cannot continue. Since this is a common
+        problem, this message expands on the issue and offers different ways to
+        solve it.
+
+        Returns:
+            string: Additional string to be appended to the default error.
+        """
+        return "This usually happens if there are redundancies in the (expanded) training data.\n" + \
+               " There are several ways to deal with this issue:\n\n" + \
+               "   - Use more data.\n\n" + \
+               "   - Add noise to the data. This can be done by slotting an additional NoiseNode\n" + \
+               "     in front of a troublesome SFANode. Noise levels do not have to be high.\n\n" + \
+               "   - Run training data through PCA. This can be done by slotting an additional\n" + \
+               "     PCA node in front of the troublesome SFANode. Use the PCA node to discard\n" + \
+               "     dimensions within your data with lower variance.\n\n" + \
+               "   - [ RECOMMENDED ] Use SVD by setting the SFANode flag use_svd to True. This\n" + \
+               "     solves the internal SFA eigenvalue problem by using two instances of singular\n" + \
+               "     value decomposition (SVD) instead of the default behavior of using the\n" + \
+               "     systems's symeig() routine to solve a single, generalized eigenvalue problem.\n" + \
+               "     Using SVD is slightly more expensive but more robust.\n"
```

Comment:
Please define this as a constant instead.

Reply:
Sorry if that's a stupid question, but Python does not have constants..?

Reply:
I think @pberkes meant to assign the string to a variable, e.g. `message="""asdasdsad"""`.

Reply:
Yes there are no real constants, but usually people use variables written in `CAPITAL_LETTERS` at the start of the module to simulate constants. This one could be called `_SINGULAR_MATRIX_ERROR_MSG`

Reply:
Right, that's an easy fix.